runtime/j9vm/java11vmi.c

```diff
@@ -1471,13 +1483,27 @@ JVM_SetBootLoaderUnnamedModule(JNIEnv *env, jobject module)
 			} else if (instanceClazz->classLoader != systemClassLoader) {
 				vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "module was not loaded by the bootclassloader");
 			} else {
+#if JAVA_SPEC_VERSION >= 21
+				J9Module *unamedModuleForSystemLoader = vm->unamedModuleForSystemLoader;
+				/* can't set bootclassloader.unnamedModule since bootclassloader hasn't finished the initialization yet */
+				if (NULL == unamedModuleForSystemLoader) {
+					vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, "unamedModuleForSystemLoader was not initialized");
+				} else if (NULL != unamedModuleForSystemLoader->moduleObject) {
+					vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, "module is already set in the unamedModuleForSystemLoader");
+				} else {
+					J9Module *j9mod = createModule(currentThread, modObj, systemClassLoader, NULL /* NULL name field */);
+					unamedModuleForSystemLoader->moduleObject = modObj;
+					Trc_MODULE_setUnamedModuleForSystemLoaderModuleObject(currentThread, j9mod, unamedModuleForSystemLoader);
+				}
+#else /* JAVA_SPEC_VERSION >= 21 */
 				if (NULL == J9VMJAVALANGCLASSLOADER_UNNAMEDMODULE(currentThread, systemClassLoader->classLoaderObject)) {
 					J9Module *j9mod = createModule(currentThread, modObj, systemClassLoader, NULL /* NULL name field */);
 					J9VMJAVALANGCLASSLOADER_SET_UNNAMEDMODULE(currentThread, systemClassLoader->classLoaderObject, modObj);
```

Comment:
Don't we still want to use `J9VMJAVALANGCLASSLOADER_SET_UNNAMEDMODULE()` for Java 21? Otherwise, how does `unnamedModule` get its value in `systemClassLoader->classLoaderObject`?

Reply:
JDK21 invokes it at L887.